### PR TITLE
#251 fileUploadDirectory - parameter that controls temporary files

### LINKED ISSUE #321: knot.x tries to access not existing /file-uploads directory
We've recently started extension development on top of knot.x to provide abstraction layer between our CMS and Solr. One of implemented endpoints is responsible for data indexing, however every POST request that carries any data ends with the following error:

```
$ curl -X POST http://localhost:8092/services/search/images/update -d "value" -v
* About to connect() to localhost port 8092 (#0)
*   Trying ::1...
* Connected to localhost (::1) port 8092 (#0)
> POST /services/search/images/update HTTP/1.1
> User-Agent: curl/7.29.0
> Host: localhost:8092
> Accept: */*
> Content-Length: 5
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 5 out of 5 bytes
< HTTP/1.1 500 java.nio.file.AccessDeniedException: /file-uploads
< Content-Type: text/plain
< Content-Length: 5552
<
Error 500: java.nio.file.AccessDeniedException: /file-uploads	at io.vertx.core.file.impl.FileSystemImpl$10.perform(FileSystemImpl.java:648)
	at io.vertx.core.file.impl.FileSystemImpl$10.perform(FileSystemImpl.java:630)
	at io.vertx.core.file.impl.FileSystemImpl.mkdirsBlocking(FileSystemImpl.java:237)
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl$BHandler.makeUploadDir(BodyHandlerImpl.java:142)
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl$BHandler.<init>(BodyHandlerImpl.java:124)
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl.handle(BodyHandlerImpl.java:65)
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl.handle(BodyHandlerImpl.java:40)
	at io.vertx.rxjava.ext.web.handler.BodyHandler.handle(BodyHandler.java:53)
	at io.vertx.rxjava.ext.web.handler.BodyHandler.handle(BodyHandler.java:34)
	at io.vertx.rxjava.ext.web.Route$1.handle(Route.java:133)
	at io.vertx.rxjava.ext.web.Route$1.handle(Route.java:131)
	at io.vertx.ext.web.impl.RouteImpl.handleContext(RouteImpl.java:217)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:78)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:118)
	at io.vertx.rxjava.ext.web.RoutingContext.next(RoutingContext.java:106)
	at io.knotx.server.SupportedMethodsAndPathsHandler.handle(SupportedMethodsAndPathsHandler.java:55)
	at io.knotx.server.SupportedMethodsAndPathsHandler.handle(SupportedMethodsAndPathsHandler.java:27)
	at io.vertx.rxjava.ext.web.Route$1.handle(Route.java:133)
	at io.vertx.rxjava.ext.web.Route$1.handle(Route.java:131)
	at io.vertx.ext.web.impl.RouteImpl.handleContext(RouteImpl.java:217)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:78)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:118)
	at io.vertx.ext.web.impl.RouterImpl.accept(RouterImpl.java:79)
	at io.vertx.rxjava.ext.web.Router.accept(Router.java:73)
	at io.vertx.rxjava.core.http.HttpServer$1.handle(HttpServer.java:90)
	at io.vertx.rxjava.core.http.HttpServer$1.handle(HttpServer.java:88)
	at io.vertx.core.http.impl.ServerConnection.handleRequest(ServerConnection.java:285)
	at io.vertx.core.http.impl.ServerConnection.processMessage(ServerConnection.java:429)
	at io.vertx.core.http.impl.ServerConnection.handleMessage(ServerConnection.java:131)
	at io.vertx.core.http.impl.HttpServerImpl$ServerHandler.lambda$createConnAndHandle$1(HttpServerImpl.java:763)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:335)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:193)
	at io.vertx.core.http.impl.HttpServerImpl$ServerHandler.createConnAndHandle(HttpServerImpl.java:757)
	at io.vertx.core.http.impl.HttpServerImpl$ServerHandler.doMessageReceived(HttpServerImpl.java:621)
	at io.vertx.core.http.impl.HttpServerImpl$ServerHandler.doMessageReceived(HttpServerImpl.java:573)
	at io.vertx.core.http.impl.VertxHttpHandler.channelRead(VertxHttpHandler.java:76)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:122)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:280)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:396)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.vertx.core.http.impl.HttpServerImpl$Http1xOrHttp2Handler.http1(HttpServerImpl.java:1064)
	at io.vertx.core.http.impl.HttpServerImpl$Http1xOrHttp2Handler.channelRead(HttpServerImpl.java:1035)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:129)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:642)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:565)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:479)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:441)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:748)
* Connection #0 to host localhost left intact
```

It seems that knot.x tries to read/write from `/file-uploads` directory that does not exist. Once I went through the following steps:

```
$ mkdir /file-uploads
$ chown knotx:knotx /file-uploads
$ ls -ld /file-uploads
drwxr-xr-x 2 knotx knotx 6 Aug  6 23:53 /file-uploads
```

and re-run the same service call it gets passed through correctly (ignore 400 error, I didn't attach valid data to my request):

```
$ curl -X POST http://localhost:8092/services/search/images/update -d "value" -v
* About to connect() to localhost port 8092 (#0)
*   Trying ::1...
* Connected to localhost (::1) port 8092 (#0)
> POST /services/search/images/update HTTP/1.1
> User-Agent: curl/7.29.0
> Host: localhost:8092
> Accept: */*
> Content-Length: 5
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 5 out of 5 bytes
< HTTP/1.1 400 Bad Request
< Content-Length: 0
<
* Connection #0 to host localhost left intact
```

Additionally I noticed that official knot.x cookbook creates empty `file-uploads` directory in `/opt/knotx`, which initially looked totally out of place to me until I encountered this issue (I raised a separate issue here: https://github.com/Knotx/knotx-cookbook/issues/3). 

After quick discussion with @tomaszmichalak  and @Skejven it seems that a global `$KNOTX_HOME` prefix may be missing and that's the reason why knot.x tries to write to `/file-uploads` instead of `/opt/knotx/<ID>/file-uploads` or `/opt/knotx/file-uploads`. 

I'm not sure what logic stands behind this `file-uploads` directory, but if that's required it should be placed in instance home dir (i.e. `/opt/knotx/<ID>/file-uploads`) to keep everything self-contained. Alternatively if this is just for data processing maybe a temporary directory should be created instead?

### PR DESCRIPTION
**Pull request by Shaihuludus**
fileUploadDirectory - new parameter in server configuration. It allows to choose where vertx should put uploaded files

Fixed #320 
Fixed #321

### CODE REVIEW

File: knotx-server/src/main/java/io/knotx/server/KnotxServerVerticle.java [hunk 1]
```diff
@@ -49,7 +51,9 @@ public void start(Future<Void> fut) throws IOException, URISyntaxException {
 
     configuration.getDefaultFlow().getEngineRouting().forEach((key, value) -> {
       if (key == HttpMethod.POST) {
-        router.route().method(key).handler(BodyHandler.create());
+        BodyHandler bodyHandler = StringUtils.isNoneEmpty(configuration.getFileUploadDirectory()) ?
```
Comment (toniedzwiedz) [inline knotx-server/src/main/java/io/knotx/server/KnotxServerVerticle.java]: `StringUtils#isNoneEmpty` implies that we're dealing with multiple values. Unless I'm reading this wrong, we should be fine with `StringUtils#isNotEmpty` or maybe `StringUtils#isNotBlank`. Not sure if it's possible to get a non-empty, blank value here.

File: knotx-server/src/main/java/io/knotx/server/configuration/KnotxServerConfiguration.java [hunk 2]
```diff
@@ -42,7 +43,7 @@ public KnotxServerConfiguration(JsonObject config) {
 
     defaultFlow = new KnotxFlowConfiguration(config.getJsonObject("defaultFlow"));
     customFlow = new KnotxFlowConfiguration(config.getJsonObject("customFlow"));
-
+    fileUploadDirectory =config.getString("fileUploadDirectory");
```
Comment (toniedzwiedz) [inline knotx-server/src/main/java/io/knotx/server/configuration/KnotxServerConfiguration.java]: Minor formatting glitch around the `=` operator.

File: knotx-server/src/main/java/io/knotx/server/KnotxServerVerticle.java [hunk 3]
```diff
@@ -80,7 +84,9 @@ public void start(Future<Void> fut) throws IOException, URISyntaxException {
     if (configuration.getCustomFlow().getEngineRouting() != null) {
       configuration.getCustomFlow().getEngineRouting().forEach((key, value) -> {
         if (key == HttpMethod.POST || key == HttpMethod.PUT || key == HttpMethod.DELETE) {
-          router.route().method(key).handler(BodyHandler.create());
+          BodyHandler bodyHandler = StringUtils.isNoneEmpty(configuration.getFileUploadDirectory()) ?
```
Comment (toniedzwiedz) [inline knotx-server/src/main/java/io/knotx/server/KnotxServerVerticle.java]: Same as above.

File: knotx-server/src/main/java/io/knotx/server/configuration/KnotxServerConfiguration.java (hunk 2 above)
Comment (Shaihuludus) [inline knotx-server/src/main/java/io/knotx/server/configuration/KnotxServerConfiguration.java]: Fixed

File: knotx-server/src/main/java/io/knotx/server/KnotxServerVerticle.java (hunk 3 above)
Comment (Shaihuludus) [inline knotx-server/src/main/java/io/knotx/server/KnotxServerVerticle.java]: fixed

File: knotx-server/src/main/java/io/knotx/server/KnotxServerVerticle.java (hunk 1 above)
Comment (Shaihuludus) [inline knotx-server/src/main/java/io/knotx/server/KnotxServerVerticle.java]: Should be "not" not "none" and notBlank will be better. It is possible to send blank value and it cause troubles in vert.x

File: knotx-server/src/main/java/io/knotx/server/configuration/KnotxServerConfiguration.java [hunk 4]
```diff
@@ -31,6 +31,7 @@
 
   private KnotxFlowConfiguration customFlow;
 
+  private String fileUploadDirectory;
```
Comment (Skejven) [inline knotx-server/src/main/java/io/knotx/server/configuration/KnotxServerConfiguration.java]: Please update [Knot.x Server documentation](https://github.com/Cognifide/knotx/wiki/Server).

**Pull request merged**